lib/index.js

```diff
@@ -2,3 +2,207 @@
+]);
+
+const REGISTRATION_DEFAULTS = {
+  grant_types: ['client_credentials'],
+  token_endpoint_auth_method: 'client_secret_post'
+};
+
+/**
+ * Returns an Express.js/Bedrock middleware function that handles OAuth 2.0
+ * Dynamic Client Registration.
+ *
+ * @see https://tools.ietf.org/html/rfc7591
+ *
+ * @param {object} options - Options hashmap.
+ * @param {object} options.authentication - Authentication strategy config,
+ *   including the endpoint authn strategy, and the validate initial token
+ *   callback. For example,
+ *   `{strategy: 'bearer', validateInitialAccessToken}`.
+ * @param {Function} options.register - Registration callback with the signature
+ *   `({registration, credentials: {initialAccessToken}})`, where the
+ *   client registration is typically persisted in a database.
+ * @param {object} [options.defaults={}] - Hashmap of default registration
+ *   values.
+ * @param {boolean} [options.allowClientProvidedId=false] - Whether to allow
+ *   clients to provide their own `client_id` values in the registration.
+ * @param {object} options.logger - Logger.
+ *
+ * @returns {Function} Returns middleware request handler.
+ */
+function clientRegistrationHandler({
+  authentication, register, defaults = {}, allowClientProvidedId = false,
+  logger
+}) {
+  return async (req, res) => {
+    try {
+      const initialAccessToken = await _authenticate({
+        req, authentication, logger});
+
+      const {body: registrationBody} = req;
+
+      const client = _parseRegistration({registrationBody, defaults});
+
+      const result = client.validateNew();
+      if(!result.valid) {
+        throw result.error;
+      }
+
+      // client can be optionally overriden by the register() callback
+      await _ensureClientId({client, allowClientProvidedId});
+      await _generateClientSecret({client, defaults});
+
+      // Epoch time
+      client.meta.client_id_issued_at = Math.floor(Date.now() / 1000);
+
+      let registrationResponse;
+      try {
+        registrationResponse = await register({
+          registration: client.meta, credentials: {initialAccessToken}
+        });
+      } catch(e) {
+        throw new InvalidRequest({
+          description: `Error in register() callback: "${e.message}".`
+        });
+      }
+
+      return _respond({registrationResponse, res});
+    } catch(error) {
+      _error({error, logger, res});
+    }
+  };
+}
+
+function _error({error, logger, res}) {
+  if(logger) {
+    logger.error('Dynamic client registration error.', {error});
+  }
+  const {
+    error: errorId, error_description: description, error_uri: uri
+  } = error;
+
+  const statusCode = error.statusCode || 400;
+  const oauth2ErrorResponse = {
+    error: errorId || 'invalid_request',
+    error_description: description || error.message,
+    error_uri: uri
+  };
+  res.set({
+    'Cache-Control': 'no-store',
+    Pragma: 'no-cache'
+  });
+  return res.status(statusCode).json(oauth2ErrorResponse);
```

Comment:
would it be possible to call next() with an error?  I think this is keeping with other middlewares like: https://github.com/digitalbazaar/bedrock-passport/blob/master/lib/index.js#L304-L307

Reply:
@mattcollier right, so, the reason this middleware doesn't do that has to do with OAuth2-dictated error formats. It can't guarantee that the next middleware in the chain will respond correctly, so, it responds itself.

Reply:
If the OAuth2 flow requires specific errors, this is ok. Thanks!